pkg/fleet-manager/pipeline/render/pipeline.go

```diff
@@ -0,0 +1,199 @@
+}
+
+// renderPipeline creates a YAML representation of the pipeline configuration.
+func renderPipeline(cfg PipelineConfig) ([]byte, error) {
+	return renderTemplate(PipelineTemplateContent, PipelineTemplateName, cfg)
+}
+
+// generateTasksInfo creates a string representation of tasks for inclusion in a pipeline.
+// It returns the name of Docker credentials if required by the tasks.
+func generateTasksInfo(pipelineName string, tasks []pipelineapi.PipelineTask) (dockerCredentials string, tasksInfo string, err error) {
+	var tasksBuilder strings.Builder
+
+	lastTask := "git-clone" // GitCloneTask is always the first task.
+	for _, task := range tasks {
+		if task.Name == "git-clone" {
+			continue // Skip the first git-clone task.because it is already fixed in template.
+		}
+
+		var taskInfo string
+		if (task.CustomTask == nil && task.PredefinedTask == nil) || (task.CustomTask != nil && task.PredefinedTask != nil) {
+			return "", "", fmt.Errorf("only one of 'PredefinedTask' or 'CustomTask' must be set in 'PipelineTask'")
+		}
+
+		if task.Name == "build-and-push-image" { // build-and-push-image need special handle.
+			taskInfo = generateKanikoTaskInfo(task.Name, generatePipelineTaskName(task.Name, pipelineName), lastTask, task.Retries)
+			dockerCredentials = DockerCredentialsName
+		} else {
+			taskInfo = generateTaskInfo(task.Name, generatePipelineTaskName(task.Name, pipelineName), lastTask, task.Retries)
+		}
+
+		fmt.Fprintf(&tasksBuilder, "  %s", taskInfo)
+		lastTask = task.Name // Update the last task.
+	}
+
+	return dockerCredentials, tasksBuilder.String(), nil
+}
+
+// generateTaskInfo formats a single task's information.
+func generateTaskInfo(taskName, taskRefer, lastTask string, retries int) string {
+	return generateTaskInfoBase(taskName, taskRefer, lastTask, retries)
+}
+
+// generateKanikoTaskInfo formats Kaniko task information, including additional Docker credentials workspace.
+func generateKanikoTaskInfo(taskName, taskRefer, lastTask string, retries int) string {
+	dockerWorkspace := "    - name: dockerconfig\n      workspace: docker-credentials\n"
+	return generateTaskInfoBase(taskName, taskRefer, lastTask, retries, dockerWorkspace)
+}
+
+// generateTaskInfoBase formats the base information of a task.
+func generateTaskInfoBase(taskName, taskRefer, lastTask string, retries int, additionalWorkspaces ...string) string {
+	var taskBuilder strings.Builder
+
+	// Define task name and reference
+	fmt.Fprintf(&taskBuilder, "- name: %s\n    taskRef:\n      name: %s\n", taskName, taskRefer)
+
+	// Specify dependency on the preceding task
+	fmt.Fprintf(&taskBuilder, "    runAfter: [\"%s\"]\n", lastTask)
+
+	// Add fixed workspace configuration
+	taskBuilder.WriteString("    workspaces:\n    - name: source\n      workspace: kurator-pipeline-shared-data\n")
+
+	// Add additional workspaces if any additionalWorkspaces exist
+	for _, workspace := range additionalWorkspaces {
+		taskBuilder.WriteString(workspace)
```

Comment:
This is hard to understand and later maintainence

Reply:
Is there a structure, otherwise I would suggest we add a tmplate

Reply:
You are right, and I had indeed considered whether to refactor this part before. But since it has been brought up, I will refactor this part of the code.